internal/xds/cache/snapshotcache.go

```diff
@@ -0,0 +1,271 @@
+
+	delete(s.streamIDNodeID, streamID)
+
+}
+
+func (s *snapshotcache) OnStreamRequest(streamID int64, req *envoy_service_discovery_v3.DiscoveryRequest) error {
+
+	nodeID := Hash.ID(req.Node)
+
+	s.streamIDNodeID[streamID] = nodeID
+
+	var nodeVersion string
+
+	var errorCode int32
+	var errorMessage string
+
+	// If no snapshot has been generated yet, we can't do anything, so don't mess with this request.
+	// go-control-plane will respond with an empty response, then send an update when a snapshot is generated.
+	if s.lastSnapshot == nil {
+		return nil
+	}
+
+	_, err := s.GetSnapshot(nodeID)
+	if err != nil {
+		err = s.SetSnapshot(context.TODO(), nodeID, s.lastSnapshot)
+		if err != nil {
+			return err
+		}
+	}
+
+	if req.Node != nil {
+		if bv := req.Node.GetUserAgentBuildVersion(); bv != nil && bv.Version != nil {
+			nodeVersion = fmt.Sprintf("v%d.%d.%d", bv.Version.MajorNumber, bv.Version.MinorNumber, bv.Version.Patch)
+		}
+	}
+
+	s.log.Debugf("Got a new request, version_info %s, response_nonce %s, nodeID %s, node_version %s", req.VersionInfo, req.ResponseNonce, nodeID, nodeVersion)
+
+	if status := req.ErrorDetail; status != nil {
+		// if Envoy rejected the last update log the details here.
+		// TODO(youngnick): Handle NACK properly
+		errorCode = status.Code
+		errorMessage = status.Message
+	}
+
+	s.log.Debugf("handling v3 xDS resource request, version_info %s, response_nonce %s, nodeID %s, node_version %s, resource_names %v, type_url %s, errorCode %d, errorMessage %s",
+		req.VersionInfo, req.ResponseNonce,
+		nodeID, nodeVersion, req.ResourceNames, req.GetTypeUrl(),
+		errorCode, errorMessage)
+
+	return nil
+}
+
+func (s *snapshotcache) OnStreamResponse(ctx context.Context, streamID int64, req *envoy_service_discovery_v3.DiscoveryRequest, resp *envoy_service_discovery_v3.DiscoveryResponse) {
+	nodeID := Hash.ID(req.Node)
+
+	s.log.Debugf("Sending Response on stream %d to node %s", streamID, nodeID)
+}
+
+// OnDeltaStreamOpen and the other OnDeltaStream*/OnStreamDelta* functions implement
+// the callbacks for the incremental xDS versions.
+// Yes, the different ordering in the name is part of the go-control-plane interface.
+func (s *snapshotcache) OnDeltaStreamOpen(ctx context.Context, streamID int64, typeURL string) error {
+
+	s.streamIDNodeID[streamID] = ""
+
+	return nil
+}
+
+func (s *snapshotcache) OnDeltaStreamClosed(streamID int64) {
+
+	delete(s.streamIDNodeID, streamID)
+
+}
+
+func (s *snapshotcache) OnStreamDeltaRequest(streamID int64, req *envoy_service_discovery_v3.DeltaDiscoveryRequest) error {
+
+	var nodeVersion string
+
+	var errorCode int32
+	var errorMessage string
+
+	nodeID := Hash.ID(req.Node)
+
+	s.streamIDNodeID[streamID] = nodeID
+
+	// If no snapshot has been generated yet, we can't do anything, so don't mess with this request.
+	// go-control-plane will respond with an empty response, then send an update when a snapshot is generated.
+	if s.lastSnapshot == nil {
```

Comment:
```suggestion
	if s.lastSnapshot != nil {
	    err := s.SetSnapshot(context.TODO(), nodeID, s.lastSnapshot); err != nil {
	     ... 
	     }  
```

Reply:
Here, I'm using a Go idiom that basically says "The leftmost side of the code should be the happy path, exceptions should be in if blocks." That way we're not indending most of the rest of the function.

Note that this is a different case to the next one - this one is that no snapshot has been generated by the rest of EG yet, the `s.GetSnapshot(nodeID)` call below checks if there's a snapshot with the current Node.